### PR TITLE
OpenCoarrays: Use MPICH over OpenMPI & add option

### PR DESCRIPTION
**Pull request by zbeekman**
- OpenMPI requires a hosts file to run on macOS which causes `brew test OpenCoarrays`
  to fail and is generally a PITA
- An option to disable failed image support which relies on
  experimental MPI features was added; a bug, fixed on MPICH master
  for quite some time but not in any release yet, causes issues in
  OpenCoarrays. This flag will help users work around that upstream
  bug.
- The OpenCoarrays development team recommends and tests against MPICH regularly
- Features like Fortran 2018 failed image support are not available through OpenMPI due to lack of support for required experimental MPI features

- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by zbeekman** [comment]
If anyone is for adding an option to enable the OpenCoarrays unit/regression test suite during installation I would like to add back at least the option to do that, even if it's not the default. I may push another commit that does this; if Homebrew maintainers disapprove it will be easy enough to reset the branch to the parent of that commit

**Comment by zbeekman** [comment]
Hold off merging this... trying to figure out what's going on with the failed image flag

**Comment by zbeekman** [comment]
OK, the issues I found should be resolved, and this PR should be ready for review.

Thanks!

File: Formula/opencoarrays.rb
Comment (fxcoudert) [inline Formula/opencoarrays.rb]: We try to limit the number of options, and run tests only for security software.

File: Formula/opencoarrays.rb
Comment (fxcoudert) [inline Formula/opencoarrays.rb]: Options mean that users have to compile from source, so we want to avoid them as much as possible. Experimental MPI support (or its absence) should be autodetected by cmake.

**Review by fxcoudert — CHANGES_REQUESTED** [review]
I am reluctant to allow the switch to mpich, because it will not allow people to install other MPI software. `openmpi` is the reference implementation of MPI in Homebrew. If we switch only opencoarrays, then other MPI-related formulas (abyss boost-mpi bsponmpi cp2k gmsh hypre konoha lammps opencoarrays scalapack sundials) cannot coexist with opencoarrays, as openmpi and mpich conflict.

**Comment by fxcoudert** [comment]
> OpenMPI requires a hosts file to run on macOS which causes brew test OpenCoarrays
to fail and is generally a PITA

```
$ brew test opencoarrays   
Testing opencoarrays
==> /usr/local/Cellar/opencoarrays/2.0.0/bin/caf tally.f90 -o tally
==> /usr/local/Cellar/opencoarrays/2.0.0/bin/cafrun -np 3 ./tally
```

Tests work here. If they don't, they need to be fixed. We can provide a host file, or arguably `cafrun` should be able to do that. Or `mpirun` can be used without a host file, if it is given the right options.

**Comment by zbeekman** [comment]
Ok let me think about this. The only way the MPICH dependency is really in
conflict is if someone wants to use coarrays + MPI since you can have both
installed and leave MPICH unlinked.

But you are right, we should probably be installing the hostfile reflecting
the machine that OC was built on and/or detect shared memory environments
at runtime.

**Pull request closed without merging**
